internal/cloud/aws/metadata.go

```diff
@@ -0,0 +1,195 @@
+	"github.com/edgelesssys/constellation/v2/internal/role"
+)
+
+const (
+	tagName = "Name"
+	tagRole = "constellation-role"
+	tagUID  = "constellation-uid"
+)
+
+type ec2API interface {
+	DescribeInstances(context.Context, *ec2.DescribeInstancesInput, ...func(*ec2.Options)) (*ec2.DescribeInstancesOutput, error)
+}
+
+type imdsAPI interface {
+	GetInstanceIdentityDocument(context.Context, *imds.GetInstanceIdentityDocumentInput, ...func(*imds.Options)) (*imds.GetInstanceIdentityDocumentOutput, error)
+	GetMetadata(context.Context, *imds.GetMetadataInput, ...func(*imds.Options)) (*imds.GetMetadataOutput, error)
+}
+
+// Metadata implements core.ProviderMetadata interface for AWS.
+type Metadata struct {
+	ec2  ec2API
+	imds imdsAPI
+}
+
+func New(cfg aws.Config) *Metadata {
+	return &Metadata{
+		ec2:  ec2.NewFromConfig(cfg),
+		imds: imds.New(imds.Options{}),
+	}
+}
+
+// Supported is used to determine if metadata API is implemented for this cloud provider.
+func (m *Metadata) Supported() bool {
+	return true
+}
+
+// List retrieves all instances belonging to the current Constellation.
+func (m *Metadata) List(ctx context.Context) ([]metadata.InstanceMetadata, error) {
+	uid, err := m.readInstanceTag(ctx, tagUID)
+	if err != nil {
+		return nil, fmt.Errorf("retrieving uid tag: %w", err)
+	}
+	ec2Instances, err := m.getAllInstancesInGroup(ctx, uid)
+	if err != nil {
+		return nil, fmt.Errorf("retrieving instances: %w", err)
+	}
+
+	return m.convertToMetadataInstance(ec2Instances)
+}
+
+// Self retrieves the current instance.
+func (m *Metadata) Self(ctx context.Context) (metadata.InstanceMetadata, error) {
+	identity, err := m.imds.GetInstanceIdentityDocument(ctx, &imds.GetInstanceIdentityDocumentInput{})
+	if err != nil {
+		return metadata.InstanceMetadata{}, fmt.Errorf("retrieving instance identity: %w", err)
+	}
+
+	name, err := m.readInstanceTag(ctx, tagName)
+	if err != nil {
+		return metadata.InstanceMetadata{}, fmt.Errorf("retrieving name tag: %w", err)
+	}
+	instanceRole, err := m.readInstanceTag(ctx, tagRole)
+	if err != nil {
+		return metadata.InstanceMetadata{}, fmt.Errorf("retrieving role tag: %w", err)
+	}
+
+	return metadata.InstanceMetadata{
+		Name:       name,
+		ProviderID: fmt.Sprintf("aws:///%s/%s", identity.AvailabilityZone, identity.InstanceID),
+		Role:       role.FromString(instanceRole),
+		VPCIP:      identity.PrivateIP,
+	}, nil
+}
+
+func (m *Metadata) getAllInstancesInGroup(ctx context.Context, uid string) ([]types.Instance, error) {
+	var instances []types.Instance
+	nextToken := (*string)(nil)
+	once := sync.Once{}
+	first := func() bool {
+		var ret bool
+		once.Do(func() { ret = true })
+		return ret
+	}
+
+	for nextToken != nil || first() {
+		describeOutput, err := m.ec2.DescribeInstances(ctx, &ec2.DescribeInstancesInput{
+			Filters: []types.Filter{
+				{
+					Name:   aws.String("tag:" + tagUID),
+					Values: []string{uid},
+				},
+			},
+			NextToken: nextToken,
+		})
```

Comment:
```suggestion

    in := &ec2.DescribeInstancesInput{
			Filters: []types.Filter{
				{
					Name:   aws.String("tag:" + tagUID),
					Values: []string{uid},
				},
			},
			NextToken: nextToken,
		}
	for out, err := m.ec2.DescribeInstances(ctx, in);; out, err = m.ec2.DescribeInstances(ctx, in) {
		...
		if describeOutput.NextToken == nil {
		    break
		}
```
Maybe something like this would be easier to understand?